File: sdks/swift/Sources/TestServer/TestServer.swift
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

struct TestServerOptions {
    let configPath: String
    let recordingDir: String
    let mode: String // "record" or "replay"
    let binaryPath: String
    let testServerSecrets: String?
}

class TestServer {
    private var process: Process?
    private let options: TestServerOptions
    
    init(options: TestServerOptions) {
        self.options = options
    }
    
    func start() async throws {
        let binaryURL: URL
        let fileManager = FileManager.default
        
        if fileManager.fileExists(atPath: options.binaryPath) {
            binaryURL = URL(fileURLWithPath: options.binaryPath)
        } else {
            let targetDir = URL(fileURLWithPath: options.binaryPath).deletingLastPathComponent()
            print("[TestServerSdk] Installing binary to \(targetDir.path)...")
            binaryURL = try await BinaryInstaller.ensureBinary(at: targetDir)
        }

        let arguments = [
            options.mode,
            "--config", options.configPath,
            "--recording-dir", options.recordingDir
        ]
        
        let process = Process()
        process.executableURL = binaryURL
        process.arguments = arguments
        
        if let secrets = options.testServerSecrets {
             var env = ProcessInfo.processInfo.environment
             env["TEST_SERVER_SECRETS"] = secrets
             process.environment = env
        }
        
        let pipe = Pipe()
        process.standardOutput = pipe
        process.standardError = pipe
        
        pipe.fileHandleForReading.readabilityHandler = { handle in
            if let data = try? handle.read(upToCount: handle.availableData.count),
               let str = String(data: data, encoding: .utf8), !str.isEmpty {
                print("[TestServer] \(str)", terminator: "")
            }
        }

        try process.run()
        self.process = process
        
        try await awaitHealthyTestServer()
    }
    
    func stop() {
        process?.terminate()
        process = nil
    }
    
    private func awaitHealthyTestServer() async throws {
        let healthURLString = try extractHealthURL(from: options.configPath)
        guard let url = URL(string: healthURLString) else {
            throw NSError(domain: "TestServer", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid health URL"])
        }
        
        print("[TestServer] Waiting for healthy server at \(url)...")
        try await checkHealth(url: url)
    }
    
    private func extractHealthURL(from configPath: String) throws -> String {
        let content = try String(contentsOfFile: configPath, encoding: .utf8)
        let fullRange = NSRange(content.startIndex..., in: content)

        // Find the first 'source_port', looks for "source_port: 1234"
        let portPattern = #"source_port:\s*(\d+)"#
        let portRegex = try NSRegularExpression(pattern: portPattern)
        let portMatch = portRegex.firstMatch(in: content, range: fullRange)
        
        guard let portRange = portMatch?.range(at: 1),
              let portRangeInString = Range(portRange, in: content) else {
            print("[TestServer] Warning: Could not parse source_port from config. Defaulting to 9000.")
            return "http://localhost:9000/health"
        }
        let port = String(content[portRangeInString])
        
        var healthPath = "/health" // Default
        let healthPattern = #"health:\s*([\w/]+)"#
        
        if let healthRegex = try? NSRegularExpression(pattern: healthPattern),
           let healthMatch = healthRegex.firstMatch(in: content, range: fullRange),
           let healthRangeInString = Range(healthMatch.range(at: 1), in: content) {
            healthPath = String(content[healthRangeInString])
        }
        
        return "http://localhost:\(port)\(healthPath)"
    }


    
    private func checkHealth(url: URL) async throws {
        let session = URLSession.shared
        let maxRetries = 20
        let delay = 0.5

        for _ in 0..<maxRetries {
            if let process = process, !process.isRunning {
                throw NSError(domain: "TestServer", code: -1, 
                            userInfo: [NSLocalizedDescriptionKey: "Server process died unexpectedly during startup."])
            }

            do {
                let (_, response) = try await session.data(from: url)
                if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                    return
                }
            } catch { /* retry */ }
            
            try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
        }
        throw NSError(domain: "TestServer", code: -1, userInfo: [NSLocalizedDescriptionKey: "Health check failed"])
    }

}

File: sdks/kotlin/src/main/kotlin/com/google/testserver/TestServer.kt
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.testserver

import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.util.concurrent.TimeUnit
import org.yaml.snakeyaml.Yaml

class TestServer(private val options: TestServerOptions) {

  private var process: Process? = null

  fun start(): Process {
    val binaryFile =
        options.binaryPath?.let { File(it) } ?: BinaryInstaller.ensureBinary(options.outDir)

    val args = mutableListOf<String>()
    args.add(binaryFile.absolutePath)
    args.add(options.mode)
    args.add("--config")
    args.add(options.configPath)
    args.add("--recording-dir")
    args.add(options.recordingDir)

    println("[TestServer] Starting test-server: ${args.joinToString(" ")}")

    val processBuilder = ProcessBuilder(args)
    processBuilder.redirectErrorStream(
        true) // Merge stdout and stderr for simplicity, or we can handle separately

    val env = processBuilder.environment()
    options.env?.forEach { (k, v) -> env[k] = v }
    options.testServerSecrets?.let { env["TEST_SERVER_SECRETS"] = it }

    val p = processBuilder.start()
    process = p

    // Log output in a separate thread
    Thread {
          p.inputStream.bufferedReader().use { reader ->
            var line = reader.readLine()
            while (line != null) {
              println("[test-server] $line")
              line = reader.readLine()
            }
          }
        }
        .start()

    // Wait for it to be healthy
    awaitHealthy()

    return p
  }

  fun stop() {
    process?.let { p ->
      if (p.isAlive) {
        println("[TestServer] Stopping test-server process (PID: ${p.pid()})...")
        p.destroy()
        if (!p.waitFor(5, TimeUnit.SECONDS)) {
          println("[TestServer] Process did not exit in time. Forcibly destroying...")
          p.destroyForcibly()
        }
        println("[TestServer] Stopped.")
      }
    }
  }

  private fun awaitHealthy() {
    val yaml = Yaml()
    val configStream = FileInputStream(options.configPath)
    val config = yaml.load<Map<String, Any>>(configStream)

    val endpoints = config["endpoints"] as? List<Map<String, Any>> ?: return

    for (endpoint in endpoints) {
      val healthPath = endpoint["health"] as? String ?: continue
      val sourceType = endpoint["source_type"] as? String ?: "http"
      val sourcePort = endpoint["source_port"]?.toString() ?: continue

      val url = "$sourceType://localhost:$sourcePort$healthPath"
      healthCheck(url)
    }
  }

  private fun healthCheck(url: String) {
    val maxRetries = 10
    var delay = 100L
    val client = HttpClient.newHttpClient()

    for (i in 0 until maxRetries) {
      try {
        val request = HttpRequest.newBuilder().uri(URI.create(url)).build()
        val response = client.send(request, HttpResponse.BodyHandlers.ofString())
        if (response.statusCode() == 200) {
          println("[TestServer] Health check passed for $url")
          return
        }
      } catch (e: Exception) {
        // Ignore and retry
        println("[TestServer] Health check attempt ${i + 1} failed for $url: ${e.message}")
      }
      Thread.sleep(delay)
      delay *= 2
    }
    throw IOException("[TestServer] Health check failed for $url after $maxRetries retries.")
  }
}

data class TestServerOptions(
    val configPath: String,
    val recordingDir: String,
    val mode: String, // "record" or "replay"
    val outDir: File = File("build/test-server"), // Where to download/resolve binary
    val binaryPath: String? = null, // Optional, if provided use this instead of downloading
    val testServerSecrets: String? = null,
    val env: Map<String, String>? = null,
)
